### PR TITLE
feat: update to use `laravel/serializable-closure` package

### PR DESCRIPTION
**Pull request by owenvoke**
This allows supporting PHP 8.1, and is required for us to continue using this package with Box, as 4.x of `opis/closure` requires FFI, which afaik isn't wanted for Box. Happy to make any changes if required. 👍🏻

It looks like Travis doesn't yet support PHP `8.1` (unless it's hardcoded to `8.1.0`) 🤷🏻

Also, looks like the quotes were changed in error messages as part of PHP 8.0 (see https://github.com/php/php-src/pull/5590), so I've updated the `testUnserializableClassStaticMethod()` test to cover this.

---

> NOTE: This can be rebased after https://github.com/amphp/parallel-functions/pull/30 and https://github.com/amphp/parallel-functions/pull/31 have been merged. 👍🏻

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -23,19 +23,19 @@
...
+        "amphp/parallel": "^1.4.1",
+        "amphp/amp": "^2.6.1",
+        "laravel/serializable-closure": "^1.0"
     },
     "require-dev": {
-        "amphp/phpunit-util": "^1.0",
-        "friendsofphp/php-cs-fixer": "^2.9",
-        "phpunit/phpunit": "^6.5"
+        "amphp/phpunit-util": "^2.0",
+        "friendsofphp/php-cs-fixer": "^3.5",
```
Comment (theofidry) [inline composer.json:33]: maybe the PHP-CS-Fixer upgrade can be moved into a separate PR?

File: composer.json [hunk 2]
```diff
@@ -23,19 +23,19 @@
...
+        "phpunit/phpunit": "^9.5.11"
```
Comment (theofidry) [inline composer.json:34]: same for PHPUnit

File: composer.json [hunk 3]
```diff
@@ -23,19 +23,19 @@
...
     },
     "config": {
         "platform": {
-            "php": "7.0.0"
+            "php": "7.4"
```
Comment (theofidry) [inline composer.json:38]: I think the bump to 7.4 and migrating to laravel/serializable-closure can be done in a separate PR

File: composer.json [hunk 4]
```diff
@@ -23,19 +23,19 @@
         }
     },
     "require": {
-        "php": ">=7",
-        "amphp/parallel": "^1.1",
-        "amphp/amp": "^2.0.3",
-        "opis/closure": "^3.0.7"
+        "php": ">=7.4",
```
Comment (theofidry) [inline composer.json:26]: I would highly recommend to switch to `^7.4 || ^8.0` (although some might say you should use `~` and specify the patch version instead)

**Review by theofidry — APPROVED** [review]

File: composer.json (hunk 4 above)
Comment (owenvoke) [reply composer.json:26]: Yeah, I also wanted to use that. 👍🏻 But was just sticking with the existing method for consistency.

File: composer.json (hunk 1 above)
Comment (owenvoke) [reply composer.json:33]: Moved to https://github.com/amphp/parallel-functions/pull/30 👍🏻

File: composer.json (hunk 2 above)
Comment (owenvoke) [reply composer.json:34]: Moved to https://github.com/amphp/parallel-functions/pull/31 👍🏻

File: composer.json (hunk 3 above)
Comment (owenvoke) [reply composer.json:38]: Updated 👍🏻

**Review by theofidry — APPROVED** [review]

File: composer.json (hunk 4 above)
Comment (kelunik) [reply composer.json]: No, I very much don't agree with that, it's just unnecessary maintenance work.

Comment (theofidry) [reply composer.json]: Increased maintenance yes but unnecessary no: you lie about the versions the package supports hence Composer will install incorrect versions requiring the user to dig into the changelogs/history to manually find a suitable tag he/she can use

Comment (kelunik) [reply composer.json]: I've spent _a lot_ more time with incorrect upper bounds than I've spent with PHP breaking changes affecting our packages. Composer's recent support for only ignoring upper bounds will improve this, but it's still not worth the time spent IMO.

Comment (owenvoke) [reply composer.json]: Do you want me to change that back to `>=7.4` then?

Comment (kelunik) [reply composer.json]: Please use >=, yes

Comment (owenvoke) [reply composer.json:26]: @kelunik, done. 👍🏻

Comment (Ocramius) [reply composer.json:26]: Generally to be avoided: PHP 9 is not gonna work OOTB anyway, especially for something as reflection-intensive as `laravel/serializable-closure`.

Should be `^7.4 || ~8.0.0 || ~8.1.0` too, since `8.2.0` is an unknown unknown too.

Comment (kelunik) [reply composer.json:26]: @Ocramius This might be true for `laravel/serializable-closure`, but not for code within this package. I know your opinion there, I just don't agree with it for our packages.

Comment (Ocramius) [reply composer.json:26]: That's fine, I'm just generally hopping around packages, explaining why this is objectively wrong, and informing other readers about it.

Comment (kelunik) [reply composer.json:26]: I know. ;-)

Comment (kelunik) [reply composer.json:26]: Sooo... even today it has cost me time: https://github.com/amphp/rpc/commit/dd4a51803869fcccd89bd0c338a8a0d857c6c6d1

File: composer.json [hunk 5]
```diff
@@ -23,10 +23,10 @@
...
+        "amphp/parallel": "^1.4.1",
+        "amphp/amp": "^2.6.1",
```
Comment (kelunik) [inline composer.json:28]: Any reason to bump these?

File: src/functions.php [hunk 6]
```diff
@@ -2,14 +2,14 @@
...
 use Amp\MultiReasonException;
 use Amp\Parallel\Sync\SerializationException;
+use function Amp\Parallel\Worker\enqueue;
 use Amp\Parallel\Worker\Pool;
 use Amp\Promise;
-use Opis\Closure\SerializableClosure;
-use function Amp\call;
-use function Amp\Parallel\Worker\enqueue;
 use function Amp\Promise\any;
+use Laravel\SerializableClosure\SerializableClosure;
```
Comment (kelunik) [inline src/functions.php:12]: The imports are misordered now, probably an issue with the new php-cs-fixer?

Comment (kelunik) [reply src/functions.php:12]: Missed that it's in the other PR, so should probably be fixed after rebase + running the fixer?

**Comment by kelunik** [comment]
Are there any compatibility concerns or is it fine to release this as v1.1.0?

**Comment by theofidry** [comment]
@kelunik AFAICS this should not introduce any API changes. I would not recommend a patch release still but bumping the minor version looks fine

**Comment by kelunik** [comment]
Yeah, we usually also increase the minor version on PHP version increments I think, independent from the dependency change.

**Comment by owenvoke** [comment]
I'll rebase this and resolve the issues mentioned hopefully tomorrow morning. 👍🏻 Cheers.

File: composer.json (hunk 5 above)
Comment (owenvoke) [reply composer.json:28]: Think that might have been accidental 🤔 I'll revert.

**Comment by owenvoke** [comment]
@kelunik / @theofidry, that should now all be sorted. 🥳 Let me know if there's anything else required. 👍🏻

**Review by kelunik — APPROVED** [review]

**Comment by kelunik** [comment]
Thanks! I really like how focused the PR is now. 😊

**Comment by owenvoke** [comment]
Yeah, much nicer to have each component update in separate PRs. 👍🏻

**Comment by owenvoke** [comment]
On a side note @kelunik, it looks like [Travis CI is no longer running on the repository](https://app.travis-ci.com/github/amphp/parallel-functions/branches). Would you be interested in me opening a PR to migrate to [GitHub Actions](https://github.com/features/actions)? Or, if not, I guess it might be worth re-enabling Travis CI. 👍🏻

**Comment by kelunik** [comment]
I know, already thought about asking you to update that as well, but just did it myself. :smile:

**Comment by kelunik** [comment]
I've tagged v1.1.0 now. :tada:

**Comment by theofidry** [comment]
Thanks a lot!

**Comment by owenvoke** [comment]
Thank you!

**Comment by leonexcc** [comment]
This breaks GrumPHP (see https://github.com/phpro/grumphp/issues/957)
